Packages: Extract Eslint config package

gziolo opened the pull request:
## Description
This PR extract `Eslint` config to its own package to make it possible to publish to npm. This also helps us to get rid of a temporary hack (introduced in #5502) of having `eslint` folder in the root directory of Gutenberg.

Related WordPress JavaScript core chat where we discussed coding standards which should be updated (https://make.wordpress.org/core/2018/7/25/javascript-chat-summary-july-24th/):

- For multiline comments, we will stick to the current core standards. Aside from some personal preferences, no one raised a strong need to change them.
- We’ll remove all spacing exceptions from the JS standards on spacing.
- Yoda conditions will no longer be enforced for JS.
- We won’t allow any exceptions for strict equality. The current exceptions in the core JS standards will be removed.
- For multiline conditions we’ll enforce three rules (proposed by @youknowriad):
  - If the condition is short enough to fit in a line, just use a single line
  - If not, one condition per line separated from the opening/closing parenthesis
  - the operator position at the end
- For multiline conditions we’ll remove the indentation clause from the standards.
- The Gutenberg naming conventions will be added to the standards. See: 
  - https://github.com/WordPress/gutenberg/pull/7670
  - https://github.com/WordPress/gutenberg/blob/4263e556fc889206fee5b22d54a1cd910f4e586b/docs/reference/coding-guidelines.md
- CSS naming conventions also need to be normalized with core. This should be addressed separately.
- We’ll start enforcing trailing comma’s.
- With regard to variable declarations at the top of each function: `var` declarations should always be at the top, `let` and `const` shouldn’t. This is because `var` is hoisted, the others aren’t.
- ES5 (non-transpiled) code should still be able to meet standards. We decided to have two rulesets, one baseset and one set specifically for ESnext (transpiled) code. It should be possible to disable the ESnext ruleset for legacy code. This will also be very useful for plugin and theme authors.
- The ESnext specific standards still need some more exploration. These will be explored in a trac ticket or Github issue / PR. There is already a related PR here: https://github.com/WordPress/gutenberg/pull/7965
- We agreed the current JS in core should be updated to adhere as much as possible to the WordPress JS coding standards. This is also in line with the most recent practices regarding coding standards. We decided to at least fix what can be autofixed.

## How has this been tested?
- `npm run lint` still passes.

## Types of changes
Refactoring - no production code has changed.

## Checklist:
- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->

gziolo committed:
Packages: Extract Eslint config package

gziolo commented on packages/eslint-config-default/package.json:
```diff
@@ -0,0 +1,29 @@
+		"eslint-config-wordpress": "^2.0.0",
+		"eslint-plugin-jsx-a11y": "^6.0.2",
+		"eslint-plugin-react": "^7.7.0",
+		"eslint-plugin-wordpress": "git://github.com/WordPress-Coding-Standards/eslint-plugin-wordpress.git#1774343f6226052a46b081e01db3fca8793cc9f1"
```
@ntwb - can we publish this plugin to npm to avoid using git commit?

youknowriad replied:
Can we move all these *-wordpress repos here?

ntwb replied:
I'll get this published once some of the coding standards ambiguities have been resolved

gziolo replied:
I copied all rules and other settings from this repository.

ntwb commented on .eslintrc.js:
```diff
@@ -19,7 +19,7 @@ const majorMinorRegExp = escapeRegExp( version.replace( /\.\d+$/, '' ) ) + '(\\.
 	root: true,
 	extends: [
-		'./eslint/config.js',
+		'@wordpress/eslint-config-default',
```
This should be `@wordpress/eslint-config-gutenberg`

youknowriad replied:
Why those configs are different? let's talk about them on meetings and align them instead of adding a "gutenberg" specific package. since we merged all the packages here, this repository is not Gutenberg specific anymore

gziolo replied:
Yes, let's talk on Tuesday about this. I'm also not happy about including Gutenberg in the name. We should also discuss what is the future plan for maintaining all the existing configurations.

gziolo replied:
I added it to the agenda: https://docs.google.com/document/d/1JjIk7I_PD7CAU7rhYEPNUza2thQVZI1RlukWEWGlYLc/edit#.

ntwb commented on docs/manifest.json:
```diff
@@ -311,6 +311,12 @@
+	{
+		"title": "@wordpress/eslint-config-default",
+		"slug": "packages-eslint-config-default",
+		"markdown_source": "https://raw.githubusercontent.com/WordPress/gutenberg/master/packages/eslint-config-default/README.md",
```
Also, each of the above should reference `eslint-config-gutenberg`

ntwb commented on package.json:
```diff
@@ -77,14 +76,14 @@
 		"@wordpress/babel-preset-default": "file:packages/babel-preset-default",
 		"@wordpress/browserslist-config": "file:packages/browserslist-config",
 		"@wordpress/custom-templated-path-webpack-plugin": "file:packages/custom-templated-path-webpack-plugin",
+		"@wordpress/eslint-config-default": "file:packages/eslint-config-default",
```
Also `@wordpress/eslint-config-gutenberg`

ntwb commented on packages/eslint-config-default/README.md:
```diff
@@ -0,0 +1,24 @@
+# @wordpress/eslint-config-default
```
Another `@wordpress/eslint-config-gutenberg`

ntwb commented on packages/eslint-config-default/README.md:
```diff
@@ -0,0 +1,24 @@
+# @wordpress/eslint-config-default
+
+Default [Eslint](https://eslint.org/) config for WordPress development.
```
This ESLint configuration is **not** for "WordPress" development, it is for "Gutenberg" development:

> `Default [Eslint](https://eslint.org/) config for Gutenberg development.`

ntwb commented on packages/eslint-config-default/package.json:
```diff
@@ -0,0 +1,29 @@
+{
+	"name": "@wordpress/eslint-config-default",
```
Again `@wordpress/eslint-config-gutenberg`

ntwb commented on packages/eslint-config-default/package.json:
```diff
@@ -0,0 +1,29 @@
+{
+	"name": "@wordpress/eslint-config-default",
+	"version": "1.0.0",
+	"description": "Default Eslint config for WordPress development",
```
`"description": "Default Eslint config for Gutenberg development",`

ntwb reviewed (dismissed):
This package is **not** the "default" ESLint config for WordPress, this package currently describes and uses the current Gutenberg coding standards, these coding standards do not reflect the current WordPress JavaScript coding standards and should not seek to represent that it is.

Side note: There are more instances of `eslint-config-default` that need to be changed to `eslint-config-gutenberg` that I did not add code comments that should also be updated.

aduth commented:
From [today's Core JS meeting](https://make.wordpress.org/core/2018/07/17/javascript-chat-summary-july-17th/), it was decided to hold off on progress here until divergences between the standards of Gutenberg and WordPress can be enumerated and reconciled.

gziolo commented:
We should also address the following comment from @aduth https://github.com/WordPress/gutenberg/pull/8132#discussion_r204424585 as part of this PR:

> Note for future reference in extracting rules: We should consider having separate rulesets for ES5 vs. ES2015+ .
> 
> Looks like this is already built-in to `eslint-plugin-wordpress`:
> 
> https://github.com/WordPress-Coding-Standards/eslint-plugin-wordpress#available-rulesets
> 
> (Confused why there's both a config and a plugin which includes rulesets)

gziolo commented:
I rebased this PR with the latest `master` changes. In addition I introduced the following changes:
- renames `@wordpress/eslint-config-default` to `@wordpress/eslint-config`
- inlined all rules for `eslint-plugin-wordpress` to avoid using git link

youknowriad commented on packages/eslint-config/package.json:
```diff
@@ -0,0 +1,28 @@
+	},
+	"dependencies": {
+		"babel-eslint": "^8.0.3",
+		"eslint-config-wordpress": "2.0.0",
```
As discussed in Core/JS chat, Let's copy this dependency here and ship.

gziolo replied:
Done with e612065.

gziolo commented:
Next steps we agreed on today during WordPress Core JavaScript weekly chat:

- publish `@wordpress/eslint-config` (as alpha)
- publish make/core post outlining proposed changes to WP standards and referencing the new published package.
- Wait a week for comments and then update the docs.

gziolo committed:
Copy config from eslint-config-wordpress

youknowriad commented on packages/eslint-config/index.js:
```diff
@@ -1,14 +1,13 @@
-		'wordpress',
-		'plugin:wordpress/esnext',
+		'./configs/wordpress.js',
+		'./configs/esnext.js',
```
Do we really need separate files here? Or maybe we should rename `wordpress.js` into `es5.js` if that's the distinction between the two?

gziolo committed:
Docs: Expose documentation for the newly added package

aduth commented on packages/eslint-config/configs/wordpress.js:
```diff
@@ -0,0 +1,103 @@
+
+module.exports = {
+	env: {
+		browser: true,
```
I think this is a good opportunity to eliminate undesirable globals which have shown to be problematic in the past, as evidenced in https://github.com/WordPress/gutenberg/pull/12070#issuecomment-439963253, https://github.com/WordPress/gutenberg/pull/9163#discussion_r211256330, and https://github.com/WordPress/gutenberg/pull/1332#issuecomment-310053886

ntwb replied:
Agreed, this has been troublesome in the past

gziolo replied:
Cleaned up in d5256b4.

gziolo committed:
Remove no longer necessary globals from the Eslint config

youknowriad commented on .eslintrc.js:
```diff
@@ -187,11 +179,25 @@ module.exports = {
 	},
 	overrides: [
+		{
+			files: [
```
What does this override do? Does it allow JEST globals? I'm fine with that personally.

gziolo replied:
I can move it back to the common settings, I wanted to ensure we don't allow accidental globals 👍

gziolo replied:
I reverted my change. It is configured again globally for Jest.

youknowriad commented on packages/block-library/src/classic/edit.js:
```diff
@@ -1,3 +1,5 @@
+/* global wp */
```
An alternative here is to use `window.wp`.

gziolo replied:
Yes, I discussed it with @aduth at the end of the day. I will update.

youknowriad reviewed (commented):
Do we already follow all the discussed guidelines? I'm surprised there's no change in or formatting.

gziolo commented:
> Do we already follow all the discussed guidelines? I'm surprised there's no change in or formatting.

It should be closely examined again, but as far as I'm aware all those changes were proposed to make coding styles used in Gutenberg to become a new standard.

gziolo commented on .eslintrc.js:
```diff
@@ -19,14 +19,8 @@ const majorMinorRegExp = escapeRegExp( version.replace( /\.\d+$/, '' ) ) + '(\\.
-		'./eslint/config.js',
-		'plugin:jest/recommended'
-	],
-	env: {
```
It turned out `plugin:jest/recommended` includes the related env and plugins config :)

gziolo commented:
/cc those who discussed at the last WordPress core JavaScript meeting: @aduth @afercia  @herregroen @nerrad @omarreiss @youknowriad

ntwb commented on packages/eslint-config/README.md:
```diff
@@ -0,0 +1,24 @@
+# Eslint Config
```
```suggestion
# ESLint Config
```

ntwb commented on packages/eslint-config/README.md:
```diff
@@ -0,0 +1,24 @@
+# Eslint Config
+
+[Eslint](https://eslint.org/) config for WordPress development.
```
```suggestion
[ESLint](https://eslint.org/) config for WordPress development.
```

ntwb commented on packages/eslint-config/package.json:
```diff
@@ -0,0 +1,27 @@
+{
+	"name": "@wordpress/eslint-config",
+	"version": "1.0.0-alpha.0",
+	"description": "Eslint config for WordPress development.",
```
```suggestion
	"description": "ESLint config for WordPress development.",
```

ntwb reviewed (approved):
A couple of inline suggested changes for ESLint spelling, otherwise 👍

ntwb committed:
Apply suggestions from code review

Fix ESLint typo
Kudos to @ntwb for catching

Co-Authored-By: gziolo <grzegorz@gziolo.pl>

gziolo commented:
It might take some time until we publish the alpha version of this package. It probably all depends when WordPress RC is released.
